File: src/main/java/org/sandboxpowered/silica/client/Silica.java
package org.sandboxpowered.silica.client;

import com.google.common.base.Joiner;
import org.apache.commons.io.FileUtils;
import org.apache.commons.io.filefilter.IOFileFilter;
import org.apache.commons.lang3.StringUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.lwjgl.glfw.GLFW;
import org.lwjgl.opengl.GL;
import org.sandboxpowered.silica.client.resources.DirectoryResourceLoader;
import org.sandboxpowered.silica.client.resources.ResourceManager;
import org.sandboxpowered.silica.client.resources.ZIPResourceLoader;
import org.sandboxpowered.silica.client.util.FileFilters;

import java.io.File;
import java.io.IOException;
import java.net.URISyntaxException;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import static org.lwjgl.glfw.GLFW.glfwSetErrorCallback;
import static org.lwjgl.opengl.GL11.*;

public class Silica implements Runnable {
    public static final Logger LOG = LogManager.getLogger(Silica.class);

    private final Window window;

    private final ResourceManager manager;


    public Silica(Args args) {
        List<String> list = new ArrayList<>();
        glfwSetErrorCallback((i, l) ->
                list.add(String.format("GLFW error during init: [0x%X]%s", i, l))
        );
        if (!GLFW.glfwInit()) {
            throw new IllegalStateException("Failed to initialize GLFW, errors: " + Joiner.on(",").join(list));
        }
        for (String string : list) {
            LOG.error("GLFW error collected during initialization: {}", string);
        }

        manager = new ResourceManager();

        try {
            URL url = Silica.class.getResource("/log4j2.xml").toURI().resolve(".").toURL();
            Path path = asPath(url);
            if (path != null) {
                if (Files.isDirectory(path)) {
                    manager.add(new DirectoryResourceLoader(path.toFile()));
                } else {
                    manager.add(new ZIPResourceLoader(path.toFile()));
                }
            }
        } catch (IOException | URISyntaxException e) {
            LOG.error("Error loading default resources", e);
        }

        File resourcePacks = new File("resourcepacks");

        if (!resourcePacks.exists())
            resourcePacks.mkdirs();

        Collection<File> files = FileUtils.listFiles(resourcePacks, FileFilters.ZIP.or(FileFilters.JAR), new IOFileFilter() {
            @Override
            public boolean accept(File file) {
                return false;
            }

            @Override
            public boolean accept(File dir, String name) {
                return false;
            }
        });
        files.forEach(file -> {
            try {
                if (file.isDirectory()) {
                    manager.add(new DirectoryResourceLoader(file));
                } else {
                    manager.add(new ZIPResourceLoader(file));
                }
            } catch (IOException e) {
                LOG.error("Failed loading resource pack {}", file.getName());
            }
        });

        LOG.debug("Loaded namespaces: [{}]", StringUtils.join(manager.getNamespaces(), ","));

        window = new Window("Sandbox Silica", args.width, args.height);

        GL.createCapabilities();

        glClearColor(0.0f, 0.0f, 0.0f, 0.0f);

        while (!window.shouldClose()) {
            glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
            window.update();
        }

        close();
    }

    public static Path asPath(URL url) {
        if (url.getProtocol().equals("file")) {
            return asFile(url).toPath();
        } else {
            try {
                return Paths.get(url.toURI());
            } catch (URISyntaxException var2) {
                LOG.error("Error finding path of url", var2);
                return null;
            }
        }
    }

    public static File asFile(URL url) {
        try {
            return new File(url.toURI());
        } catch (URISyntaxException var2) {
            LOG.error("Error finding file of url", var2);
            return null;
        }
    }

    public void close() {
        window.cleanup();
    }

    @Override
    public void run() {

    }

    public static class Args {
        public final int width, height;

        public Args(int width, int height) {
            this.width = width;
            this.height = height;
        }
    }
}

File: client/src/main/kotlin/org/sandboxpowered/silica/client/Silica.kt
package org.sandboxpowered.silica.client

import com.google.common.base.Joiner
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import org.lwjgl.glfw.GLFW
import org.lwjgl.system.Configuration
import org.sandboxpowered.silica.resources.*
import org.sandboxpowered.silica.util.*
import org.sandboxpowered.silica.util.extensions.join
import org.sandboxpowered.silica.util.extensions.listFiles
import org.sandboxpowered.silica.util.extensions.notExists
import java.io.File
import java.io.IOException
import java.nio.file.Path
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicInteger
import kotlin.math.max


class Silica(private val args: Args) : Runnable {
    private val logger: Logger = LogManager.getLogger()
    lateinit var window: Window
    private lateinit var assetManager: ResourceManager
    lateinit var renderer: Renderer

    private fun close() {
        window.cleanup()
    }

    private val DEBUG: Boolean = Configuration.DEBUG.get(false)
    private val executorService = Executors.newFixedThreadPool(
        max(1, Runtime.getRuntime().availableProcessors() / 2)
    ) { r: Runnable? ->
        val t = Thread(r)
        t.priority = Thread.MIN_PRIORITY
        t.name = "Chunk builder"
        t.isDaemon = true
        t
    }
    private val chunkBuildTasksCount = AtomicInteger()
    private val updateAndRenderRunnables: Queue<DelayedRunnable> = ConcurrentLinkedQueue()

    private class DelayedRunnable(val runnable: Runnable, val name: String, val delay: Int) {
    }

    init {
        if (DEBUG) {
            /* When we are in debug mode, enable all LWJGL debug flags */
            System.setProperty("org.lwjgl.util.Debug", "true")
            System.setProperty("org.lwjgl.util.NoChecks", "false")
            System.setProperty("org.lwjgl.util.DebugLoader", "true")
            System.setProperty("org.lwjgl.util.DebugAllocator", "true")
            System.setProperty("org.lwjgl.util.DebugStack", "true")
        }
        /* Configure LWJGL MemoryStack to 1024KB */
        Configuration.STACK_SIZE.set(1024)
    }

    override fun run() {
        if (init())
            return
        while (!window.shouldClose()) {
            renderer.frame()
            window.update()
        }
        close()
    }

    data class Args(val width: Int, val height: Int, val renderer: String, val minecraftPath: Path?)

    class InvalidRendererException(message: String) : RuntimeException(message)

    private fun init(): Boolean {
        val serviceLoader = ServiceLoader.load(RenderingFactory::class.java)
        val renderers = serviceLoader.toList()

        renderer = when {
            args.renderer.isNotEmpty() -> {
                val factory = renderers.find { it.getId() == args.renderer }
                factory?.createRenderer(this)
                    ?: throw InvalidRendererException("${args.renderer} renderer is not supported")
            }
            renderers.isEmpty() -> throw UnknownError("No renderers defined")
            renderers.size == 1 -> renderers[0].createRenderer(this)
            else -> {
                val sorted = renderers.sortedBy { it.getPriority() }
                sorted[0].createRenderer(this)
            }
        }
        logger.debug("Using Renderer: ${renderer.getName()}")
        val list: MutableList<String?> = ArrayList()
        GLFW.glfwSetErrorCallback { i: Int, l: Long ->
            list.add(String.format("GLFW error during init: [0x%X]%s", i, l))
        }
        check(GLFW.glfwInit()) { "Failed to initialize GLFW, errors: " + Joiner.on(",").join(list) }
        for (string in list) {
            logger.error("GLFW error collected during initialization: {}", string)
        }
        if (list.isNotEmpty())
            return true
        assetManager = ResourceManager(ResourceType.ASSETS)
        assetManager.add(ClasspathResourceLoader())
        Util.findMinecraft(assetManager, args.minecraftPath)

        File("resourcepacks").apply {
            if (notExists()) {
                mkdirs()
            } else if (isFile) {
                delete()
                mkdirs()
            }
        }.listFiles(FileFilters.ZIP) { file ->
            try {
                if (file.isDirectory) {
                    assetManager.add(DirectoryResourceLoader(file))
                } else {
                    assetManager.add(ZIPResourceLoader(file))
                }
            } catch (e: IOException) {
                logger.error("Failed loading resource source {}", file.name)
            }
        }

        logger.debug("Loaded namespaces: [${assetManager.getNamespaces().join(",")}]")
        window = Window("Sandbox Silica", args.width, args.height, renderer)
        renderer.init()
        return false
    }
}